include/ozo/binary_serialization.h

```diff
@@ -0,0 +1,138 @@
+    hana::for_each(
+        hana::to<hana::tuple_tag>(hana::make_range(hana::size_c<0>, hana::size_c<sizeof(T)>)),
+        [&] (auto i) { *out++ = value >> decltype(i)::value * CHAR_BIT & std::numeric_limits<std::uint8_t>::max(); }
+    );
+}
+
+template <class T>
+struct floating_point_integral {};
+
+template <>
+struct floating_point_integral<float> {
+    using type = std::uint32_t;
+};
+
+template <>
+struct floating_point_integral<double> {
+    using type = std::uint64_t;
+};
+
+template <class T>
+using floating_point_integral_t = typename floating_point_integral<T>::type;
+
+template <class OutIteratorT, class T>
+constexpr std::enable_if_t<std::is_floating_point_v<T>, void> write(T value, OutIteratorT out) {
+    union {
+        T input;
+        floating_point_integral_t<T> output;
+    } data {value};
+    write(data.output, out);
+}
+
+template <class T>
+constexpr std::size_t get_dimension(const T&) noexcept {
+    return 1;
+}
+
+template <class T>
+constexpr std::enable_if_t<std::is_arithmetic_v<T>, std::size_t> get_size(const T&) noexcept {
+    return sizeof(T);
+}
+
+template <class T>
+constexpr std::enable_if_t<!std::is_arithmetic_v<T>, std::size_t> get_size(const T& value) noexcept {
+    return std::size(value);
+}
+
+} // namespace detail
+
+template <class T>
+using single_byte_integral_t = std::enable_if_t<std::is_integral_v<T> && sizeof(T) == 1, void>;
+
+template <class T, class OidMapT, class OutIteratorT>
+constexpr single_byte_integral_t<T> send(T value, const OidMapT&, OutIteratorT out) {
+    detail::write(value, out);
+}
+
+template <class T>
+using multi_byte_integral_t = std::enable_if_t<std::is_integral_v<T> && (sizeof(T) == 2 || sizeof(T) == 4 || sizeof(T) == 8), void>;
+
+template <class T, class OidMapT, class OutIteratorT>
+constexpr multi_byte_integral_t<T> send(T value, const OidMapT&, OutIteratorT out) {
+    detail::write(detail::bytes_order_swap(value), out);
+}
+
+template <class T>
+using floating_point_t = std::enable_if_t<std::is_floating_point_v<T> && (sizeof(T) == 4 || sizeof(T) == 8), void>;
+
+template <class T, class OidMapT, class OutIteratorT>
+constexpr floating_point_t<T> send(T value, const OidMapT&, OutIteratorT out) {
+    detail::write(value, out);
+}
+
+template <class OidMapT, class OutIteratorT, class CharT, class TraitsT, class AllocatorT>
+constexpr void send(const std::basic_string<CharT, TraitsT, AllocatorT>& value, const OidMapT&, OutIteratorT out) {
+    std::for_each(value.begin(), value.end(), [&] (auto v) { detail::write(v, out); });
+}
+
+template <class OidMapT, class OutIteratorT, class T, class AllocatorT>
+constexpr void send(const std::vector<T, AllocatorT>& value, const OidMapT& oid_map, OutIteratorT out) {
+    send(std::int32_t(detail::get_dimension(value)), oid_map, out);
+    send(std::int32_t(0), oid_map, out);
+    send(type_oid<T>(oid_map), oid_map, out);
+    send(std::int32_t(std::size(value)), oid_map, out);
+    send(std::int32_t(1), oid_map, out);
```

Comment:
Why should I do this? For what reason we need to increase compexity of code to write simple sequence of functions calls? I don't see any profit.

Reply:
Well, if you need to change a function signature or something like this you will do it once. And more generally, if you need to write different headers for different types you can use traits instead of different functions.

BTW, I was wrong about boilerplate code.
http://www.boost.org/doc/libs/1_61_0/libs/hana/doc/html/group__group-Struct.html#gad301dd8e9fb4639d7874619c97d6d427
hana::members can return a tuple of values of an adapted structure. In this case, if you define the header structure it will be easy to understand what the header is.

Reply:
Done